sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubProducerClient.cs

```diff
@@ -1361,7 +1356,7 @@ private static void AssertPartitionIsReferenced(string partitionId)
         private static void AssertIdempotentBatchNotPublished(EventDataBatch batch)
         {
             if ((batch.StartingPublishedSequenceNumber.HasValue)
-                || (batch.AsEnumerable<EventData>().Any(eventData => eventData.PublishedSequenceNumber.HasValue)))
+                || (batch.AsList<EventData>().Any(eventData => eventData.PublishedSequenceNumber.HasValue)))
```

Comment:
To my knowledge, the `Any` call here is non-allocating and is [implemented](https://source.dot.net/#System.Linq/System/Linq/AnyAll.cs,47) the same way that we would unroll it other than two `if` checks.